std/experimental/ndslice/package.d

```diff
@@ -0,0 +1,519 @@
+        .array               // 9. C[] : sole memory allocation in this function
+        .sliced(wnds.shape); //10. 3D : returns slice with corresponding shape
+}
+---
+
+A function that calculates the value of iterator median is also necessary.
+
+---
+/++
+Params:
+    r = input range
+    buf = buffer with length no less than the number of elements in `r`
+Returns:
+    median value over the range `r`
++/
+T median(Range, T)(Range r, T[] buf)
+{
+    import std.algorithm.sorting: sort;
+    size_t n;
+    foreach (e; r)
+        buf[n++] = e;
+    buf[0 .. n].sort();
+    immutable m = n >> 1;
+    return n & 1 ? buf[m] : cast(T)((buf[m - 1] + buf[m]) / 2);
+}
+---
+
+The `main` function:
+---
+void main(string[] args)
+{
+    import std.conv: to;
+    import std.getopt: getopt, defaultGetoptPrinter;
+    import std.path: stripExtension;
+
+    uint nr, nc, def = 3;
+    auto helpInformation = args.getopt(
+        "nr", "number of rows in window, default value is " ~ def.to!string, &nr,
+        "nc", "number of columns in window default value equals to nr", &nc);
+    if (helpInformation.helpWanted)
+    {
+        defaultGetoptPrinter(
+            "Usage: median-filter [<options...>] [<file_names...>]\noptions:",
+            helpInformation.options);
+        return;
+    }
+    if (!nr) nr = def;
+    if (!nc) nc = nr;
+
+    auto buf = new ubyte[nr * nc];
+
+    foreach (name; args[1 .. $])
+    {
+        import imageformats; // can be found at code.dlang.org
+
+        IFImage image = read_image(name);
+
+        auto ret = image.pixels
+            .sliced(image.h, image.w, image.c)
+            .movingWindowByChannel
+                !(window => median(window.byElement, buf))
+                 (nr, nc);
+
+        write_image(
+            name.stripExtension ~ "_filtered.png",
+            ret.length!1,
+            ret.length!0,
+            (&ret[0, 0, 0])[0 .. ret.elementsCount]);
+    }
+}
+---
+
+This program works both with color and grayscale images.
+
+---
+$ median-filter --help
+Usage: median-filter [<options...>] [<file_names...>]
+options:
+     --nr number of rows in window, default value is 3
+     --nc number of columns in window default value equals to nr
+-h --help This help information.
+---
+
+$(H2 Compared with `numpy.ndarray`)
+
+`numpy` is undoubtedly one of the most effective software packages that has
+facilitated the work of many engineers and scientists. However, due to the
+specifics of implementation of Python, a programmer who wishes to use the
+functions not represented in `numpy` may find that the built-in functions
+implemented specifically for `numpy` are not enough, and their Python
+implementations work at a very low speed. An extension of `numpy` might be a
+solution in this case; nevertheless, it does not seem to be practical because
+even the most basic `numpy` functions that refer directly to `ndarray`
+data must be implemented in C to be productive enough.
```

Comment:
Suggest:
Extending numpy can be done, but is somewhat laborious as even the most basic `numpy` functions that refer directly to `ndarray` data must be implemented in C for reasonable performance.